File: data-ext/download_scopus_citations.R

# depends on ds_redline object
if (!"ds_redline" %in% ls()) {
  source("R/results_query.R")
}

# depend on p_values object
if (!"p_values" %in% ls()) {
  source("R/results_pvalues.R")
}

## ---- publications ----

pubs <- readRDS("output/publications.rds")
# jif <- read_csv("data/JIF_incites.csv", skip = 1)

# Drop duplicate Id column
pubs <- pubs %>% select(-Id)

pvals_pub <- p_values %>% 
  select(Accession, suppdata_id, pi0)

pubs <- ds_redline %>% 
  select(Accession, PubMedIds, model) %>% 
  inner_join(pubs) %>% 
  mutate(ISSN = case_when(str_length(ISSN) == 0 ~ ESSN,
                          str_length(ISSN) != 0 ~ ISSN))


# Download citations ------------------------------------------------------

dois <- pubs %>% 
  filter(str_detect(model, "Human")) %>% 
  select(Accession, PubMedIds, DOI) %>% 
  distinct() %>% 
  nest(Accession)

scopus <- crul::HttpClient$new(url = 'https://api.elsevier.com', 
                    headers = list(Accept = "application/json",
                                   Encoding = "UTF-8"))

#' Query function
scopus_n_citations <- function(doi, con) { 
  con$get(path = "content/search/scopus", 
          query = list(query = glue::glue("doi({doi})"),
                       apiKey = Sys.getenv("Elsevier_API")))
}

#' Run query
library(jsonlite)
dois <- dois %>% 
  mutate(resp = map(DOI, scopus_n_citations, con = scopus))
dois <- dois %>% 
  filter(map_lgl(resp, ~ .x$status_code == 200)) %>% 
  mutate(parsed = map(resp, ~ .x$parse()),
         json = map(parsed, jsonlite::fromJSON))

#' Extract citations 
extr_citations <- function(x) {
  citedbycount <- x$`search-results`$entry$`citedby-count`
  if (is.null(citedbycount)) return(NA)
  as.numeric(citedbycount[1])
}

#' Select results for saving
scopcitations <- dois %>% 
  mutate(citations = map_dbl(json, extr_citations)) %>% 
  select(PubMedIds, DOI, citations, json)

#' Save results to rds
readr::write_rds(scopcitations, "data/scopus_citedbycount.rds")

File: scripts/preprocess/download_scopus_citations.py
import pandas as pd
import requests

api_key = snakemake.params.get("api_key", None)
if api_key is None:
    raise Exception("API key is required for request")

# Import document summaries
pubs = pd.read_csv(snakemake.input[0], sep=",")

# Get list of dois
dois = pubs["DOI"].tolist()

# Query parameters
url = "https://api.elsevier.com/content/search/scopus"
headers = {"Accept": "application/json", "Encoding": "UTF-8"}


# Run query
with open(snakemake.output[0], "a") as f:
    citations = pd.DataFrame(columns=["DOI", "citations"])
    for doi in dois:
        params = {"query": "doi({})".format(doi), "apiKey": api_key}
        r = requests.get(url, headers=headers, params=params)
        entry = r.json()["search-results"]["entry"][0]
        try:
            entry = {"DOI": doi, "citations": entry["citedby-count"]}
        except KeyError:
            entry = {"DOI": doi, "citations": "NA"}
        citations = citations.append(
            {k: v for k, v in entry.items()}, ignore_index=True
        )
    citations.to_csv(f, mode="a", header=not f.tell(), index=False)
